common/raw_hashtable.h

```diff
@@ -0,0 +1,1206 @@
+  template <ssize_t OtherSmallSize>
+  explicit TableImpl(const TableImpl<BaseT, OtherSmallSize>& arg)
+      : TableImpl() {
+    this->CopyFrom(arg);
+  }
+  TableImpl(TableImpl&& arg) noexcept : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+  template <ssize_t OtherSmallSize>
+  explicit TableImpl(TableImpl<BaseT, OtherSmallSize>&& arg) : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+
+  auto ResetImpl() -> void;
+
+ private:
+  using KeyT = BaseT::KeyT;
+  using ValueT = BaseT::ValueT;
+
+  // Do a bunch of validation of the small size to establish our invariants.
+  // Using `static_assert` instead of a `requires` clause lets us associate a
+  // message with each constraint.
+  static_assert(llvm::isPowerOf2_64(SmallSize),
+                "SmallSize must be a power of two for a hashed buffer!");
+  static_assert(SmallSize >= MaxGroupSize,
+                "We require all small sizes to multiples of the largest group "
+                "size supported to ensure it can be used portably.  ");
+  static_assert((SmallSize % MaxGroupSize) == 0,
+                "Small size must be a multiple of the max group size supported "
+                "so that we can allocate a whole number of groups.");
+  // Implied by the max asserts above.
+  static_assert(SmallSize >= GroupSize);
+  static_assert((SmallSize % GroupSize) == 0);
+
+  static_assert(SmallSize >= alignof(StorageEntry<KeyT, ValueT>),
+                "Requested a small size that would require padding between "
+                "metadata bytes and correctly aligned key and value types. "
+                "Either a larger small size or a zero small size and heap "
+                "allocation are required for this key and value type.");
+
+  // A concrete implementation of storage for the provided key type, value type,
+  // and small size.
+  struct SmallStorage : Storage {
+    alignas(BaseT::Alignment) uint8_t metadata[SmallSize];
+    mutable StorageEntry<KeyT, ValueT> entries[SmallSize];
+  };
+
+  auto small_storage() const -> RawHashtable::Storage* {
+    return &small_storage_;
+  }
+
+  mutable SmallStorage small_storage_;
+};
+
+// A template specialization for when no small size optimization buffer is
+// desired.
+template <typename InputBaseT>
+class TableImpl<InputBaseT, 0> : public InputBaseT {
+ protected:
+  using BaseT = InputBaseT;
+
+  TableImpl() : BaseT(0, nullptr) {}
+  TableImpl(const TableImpl& arg) : TableImpl() { this->CopyFrom(arg); }
+  template <ssize_t OtherMinSmallSize>
+  explicit TableImpl(const TableImpl<BaseT, OtherMinSmallSize>& arg)
+      : TableImpl() {
+    this->CopyFrom(arg);
+  }
+  TableImpl(TableImpl&& arg) noexcept : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+  template <ssize_t OtherMinSmallSize>
+  explicit TableImpl(TableImpl<BaseT, OtherMinSmallSize>&& arg) : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+
+  auto ResetImpl() -> void;
+};
+
```

Comment:
Some sort of divider to separate implementation details from things used by clients of this library?